File: algorithms/0872.leaf-similar-trees/solution.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    pub fn leaf_similar(
        root1: Option<Rc<RefCell<TreeNode>>>,
        root2: Option<Rc<RefCell<TreeNode>>>,
    ) -> bool {
        let mut seq_left: Vec<i32> = vec![];
        let mut seq_right: Vec<i32> = vec![];

        Solution::leaf_sequence(root1.as_ref(), &mut seq_left);
        Solution::leaf_sequence(root2.as_ref(), &mut seq_right);

        seq_left == seq_right
    }

    fn leaf_sequence(root: Option<&Rc<RefCell<TreeNode>>>, current: &mut Vec<i32>) {
        if let Some(node) = root {
            let node = node.borrow();

            if node.left.is_none() && node.right.is_none() {
                current.push(node.val);
            }

            Solution::leaf_sequence(node.left.as_ref(), current);
            Solution::leaf_sequence(node.right.as_ref(), current);
        }
    }
}

File: algorithms/0872.leaf-similar-trees/solution.cpp
#include <vector>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        vector<int> seqLeft;
        vector<int> seqRight;

        leafSequence(root1, seqLeft);
        leafSequence(root2, seqRight);

        return seqLeft == seqRight;
    }

private:
    void leafSequence(TreeNode* root, vector<int>& current) {
        if (root) {
            if (root->left == nullptr && root->right == nullptr) {
                current.push_back(root->val);
            }
            leafSequence(root->left, current);
            leafSequence(root->right, current);
        }
    }
};
